File: src/main/java/xyz/oribuin/lilori/commands/moderation/CmdPurge.java
package xyz.oribuin.lilori.commands.moderation;

import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.ChannelType;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import xyz.oribuin.lilori.managers.command.Command;
import xyz.oribuin.lilori.managers.command.CommandEvent;
import xyz.oribuin.lilori.utils.EventWaiter;

import java.time.OffsetDateTime;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class CmdPurge extends Command {

    private EventWaiter waiter;

    public CmdPurge(EventWaiter waiter) {
        this.name = "Purge";
        this.aliases = Collections.singletonList("clear");
        this.description = "Mass clear server messages.";
        //this.arguments = "<Channel/Msgs/User> <#Channel/Number/@User>";
        this.botPermissions = new Permission[]{Permission.MESSAGE_MANAGE, Permission.MANAGE_PERMISSIONS, Permission.MANAGE_CHANNEL};
        this.userPermissions = new Permission[]{Permission.MESSAGE_MANAGE, Permission.MANAGE_PERMISSIONS, Permission.MANAGE_CHANNEL};
        this.waiter = waiter;
    }

    @Override
    public void executeCommand(CommandEvent event) {

        // Variables
        String[] args = event.getMessage().getContentRaw().split(" ");
        TextChannel channel = (TextChannel) event.getChannel();
        OffsetDateTime time = event.getMessage().getTimeCreated();
        if (event.getMember() == null) return;

        if (args.length <= 2) {
            event.deleteCmd(10, TimeUnit.SECONDS);
            event.timedReply(event.getAuthor().getAsMention() + ", Correct Format: ;purge " + this.getArguments(), 10, TimeUnit.SECONDS);
            return;
        }

        if (args[1].equalsIgnoreCase("messages") || args[1].equalsIgnoreCase("msgs")) {
            if (args[2] == null) {
                event.deleteCmd(10, TimeUnit.SECONDS);
                event.timedReply(event.getAuthor().getAsMention() + ", Please include the number of messages to delete. (Between 2-100)", 10, TimeUnit.SECONDS);
                return;
            }

            try {
                int msgCount = Integer.parseInt(args[2]);

                if (msgCount <= 1 || msgCount > 100) {
                    event.timedReply(event.getAuthor().getAsMention() + ", You can only purge up to 100 messages at once, You provided **" + msgCount + "**.", 10, TimeUnit.SECONDS);
                    event.deleteCmd(10, TimeUnit.SECONDS);
                    return;
                }

                event.deleteCmd();
                List<Message> msgs = channel.getHistory().retrievePast(msgCount).complete();

                if (msgs.size() > 1)
                    event.getChannel().purgeMessages(msgs);

            } catch (NumberFormatException e) {
                event.deleteCmd(10, TimeUnit.SECONDS);
                event.timedReply(event.getAuthor().getAsMention() + ", Please include valid arguments.", 10, TimeUnit.SECONDS);
            }
        } else if (args[1].equalsIgnoreCase("channel")) {

            if (event.getMessage().getMentionedChannels().size() == 0) {
                event.timedReply(event.getAuthor().getAsMention() + ", Please mention a text channel to purge.", 10, TimeUnit.SECONDS);
                event.deleteCmd(10, TimeUnit.SECONDS);
                return;
            }

            if (event.getMessage().getMentionedChannels().get(0).getType() != ChannelType.TEXT) {
                event.timedReply(event.getAuthor().getAsMention() + ", Please include a text channel to purge.", 10, TimeUnit.SECONDS);
                event.deleteCmd(10, TimeUnit.SECONDS);
                return;
            }

            if (args[2].equalsIgnoreCase(event.getMessage().getMentionedChannels().get(0).getAsMention())) {

                TextChannel textChannel = event.getMessage().getMentionedChannels().get(0);

                if (textChannel == null) {
                    event.timedReply(event.getAuthor().getAsMention() + ", textChannel is null...", 10, TimeUnit.SECONDS);
                    event.deleteCmd(10, TimeUnit.SECONDS);
                    return;
                }

                if (event.getMember() != null)
                    if (!event.getMember().hasPermission(textChannel)) {
                        event.timedReply(event.getAuthor().getAsMention() + ", You cannot purge this channel.", 10, TimeUnit.SECONDS);
                        event.deleteCmd(10, TimeUnit.SECONDS);
                        return;
                    }

                event.timedReply(event.getAuthor().getAsMention() + ", You are about to purge " + textChannel.getAsMention() + ", Please type **confirm** to continue. (Note: This deletes existing webhooks)", 2, TimeUnit.MINUTES);
                event.deleteCmd(10, TimeUnit.SECONDS);

                String getWeek = time.getDayOfWeek().name().substring(0, 1).toUpperCase() + time.getDayOfWeek().name().substring(1).toLowerCase();
                waiter.waitForEvent(GuildMessageReceivedEvent.class,
                        predicate -> predicate.getAuthor().equals(event.getMessage().getAuthor())
                                && predicate.getMessage().getContentRaw().toLowerCase().contains("confirm")
                                || predicate.getMessage().getContentRaw().toLowerCase().contains("cancel"),
                        consumer -> {
                            if (consumer.getMessage().getContentRaw().toLowerCase().contains("confirm")) {

                                EmbedBuilder embedBuilder = new EmbedBuilder()
                                        .setAuthor("Successfully Purged Channel")
                                        .setDescription("**Channel:** #" + textChannel.getName() + "\n" +
                                                "**Purged By:** " + event.getAuthor().getAsMention() + "\n" +
                                                "**Purged on:** " + getWeek + " at " + time.getHour() + ":" + time.getMinute());

                                consumer.getMessage().delete().queueAfter(2, TimeUnit.MINUTES);
                                event.getGuild().createTextChannel(textChannel.getName())
                                        .setNSFW(textChannel.isNSFW())
                                        .setSlowmode(textChannel.getSlowmode())
                                        .setTopic(textChannel.getTopic())
                                        .setParent(textChannel.getParent())
                                        .setPosition(textChannel.getPosition()).queue(getChannel -> {

                                    textChannel.getPermissionOverrides().forEach(permissionOverride -> getChannel.createPermissionOverride(permissionOverride.getPermissionHolder()).queue());
                                    getChannel.sendMessage(event.getAuthor().getAsMention()).queue();
                                    getChannel.sendMessage(embedBuilder.build()).queue();
                                });
                                textChannel.delete().queue();
                            } else {
                                waiter.shutdown();
                                consumer.getMessage().addReaction("✅").queue();
                            }
                        }, 2, TimeUnit.MINUTES, () -> {
                            if (waiter.isShutdown())
                                waiter.shutdown();
                        });
            }
        } else if (args[1].equalsIgnoreCase("user")) {

            if (event.getMessage().getMentionedMembers().size() == 0) {
                event.timedReply(event.getAuthor().getAsMention() + ", Please include a user to purge.", 10, TimeUnit.SECONDS);
                event.deleteCmd(10, TimeUnit.SECONDS);
                return;
            }

            Member member = event.getMessage().getMentionedMembers().get(0);
            // || member.getUser().isBot() || event.isHigher(member, event.getMember())
            if (member.hasPermission(Permission.ADMINISTRATOR)) {
                event.timedReply(event.getMessage().getAuthor().getAsMention() + ", You cannot purge this user's messages.", 10, TimeUnit.SECONDS);
                event.deleteCmd(10, TimeUnit.SECONDS);
                return;
            }

            event.getGuild().getTextChannels().forEach(textChannel -> {
                for (Message message : textChannel.getHistory().retrievePast(100).complete()) {
                    if (message.getAuthor().equals(member.getUser())) {
                        message.delete().queue();
                    }
                }
            });
        } else {
            event.deleteCmd(10, TimeUnit.SECONDS);
            event.timedReply(event.getAuthor().getAsMention() + ", Correct Format: ;purge [Msgs/User/Channel] [Number/#Channel/@User]", 10, TimeUnit.SECONDS);
        }
    }
}

File: src/main/kotlin/xyz/oribuin/lilori/command/moderation/CmdPurge.kt
package xyz.oribuin.lilori.command.moderation

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.events.message.guild.react.GuildMessageReactionAddEvent
import org.apache.commons.lang3.StringUtils
import xyz.oribuin.lilori.LilOri
import xyz.oribuin.lilori.handler.Category
import xyz.oribuin.lilori.handler.Cmd
import xyz.oribuin.lilori.handler.Command
import xyz.oribuin.lilori.handler.CommandEvent
import xyz.oribuin.lilori.util.BotUtils
import xyz.oribuin.lilori.util.EventWaiter

@Cmd(
        name = "Purge",
        description = "Purge messages from the server.",
        category = Category.Type.MODERATION,
        arguments = ["<msg/channel>", "<count/#channel>"],
        aliases = [],
        userPermissions = [Permission.MESSAGE_MANAGE, Permission.MANAGE_PERMISSIONS, Permission.MANAGE_CHANNEL],
        botPermissions = [Permission.MESSAGE_MANAGE, Permission.MANAGE_PERMISSIONS, Permission.MANAGE_CHANNEL],
        guildId = ""
)
class CmdPurge(bot: LilOri, private val waiter: EventWaiter) : Command(bot) {

    // ignore all of this
    override fun executeCommand(event: CommandEvent) {
        if (event.args.size <= 2) {
            event.sendEmbedReply("❗ Invalid Arguments", "The correct usage is ${event.prefix}${getAnnotation(javaClass).name.toLowerCase()} ${BotUtils.formatList(getAnnotation(javaClass).arguments.toList())}")
            return
        }

        when (event.args[1].toLowerCase()) {
            "messages", "msgs" -> {
                try {
                    // Define the message count
                    val messageCount = Integer.parseInt(event.args[2])

                    // Check the message count
                    if (messageCount <= 1 || messageCount > 100) {
                        event.sendEmbedReply("❗ Invalid Number", "You can only purge 2-100 messages due to API Limitations")
                        return
                    }

                    // Delete command
                    event.deleteCmd()

                    // Define the list of messages
                    val messages = event.channel.history.retrievePast(messageCount).complete()

                    // If the message size
                    event.channel.purgeMessages(messages)

                } catch (ex: NumberFormatException) {
                    event.sendEmbedReply("❗ Invalid Arguments", "The correct usage is ${event.prefix}${getAnnotation(javaClass).name.toLowerCase()} ${BotUtils.formatList(getAnnotation(javaClass).arguments.toList())}")
                }
            }

            "channel" -> {
                // Check if a channel was mentioned
                if (event.message.mentionedChannels.size == 0) {
                    event.sendEmbedReply("❗ Invalid Channel", "Please mention a valid channel.")
                    return
                }

                // Define the channel mentioned
                val channel = event.message.mentionedChannels[0]

                // Check if the channel is a text channel
                if (channel !is TextChannel) {
                    event.sendEmbedReply("❗ Invalid Channel Type", "You can only purge text channels!")
                    return
                }

                if (!event.member.hasPermission(channel) || event.selfMember.hasPermission(channel)) {
                    event.sendEmbedReply("❗ Can't Purge", "I cannot purge that channel due to no permission")
                    return
                }

                // Send the starting confirmation message
                event.channel.sendMessage(event.author.asMention + ", You are about to purge (delete) ${channel.asMention}, Please react with ✅ to continue, React with ❌ to cancel the purge. (Note: __This deletes existing webhooks__)").queue { msg ->
                    // Add the message reactions
                    msg.addReaction("✅").queue()
                    msg.addReaction("❌").queue()

                    // Define the time the message was created
                    val time = event.message.timeCreated

                    // Wait for the GuildMessageReactionAddEvent, Check if the user is the author and the message is the confirmation message
                    waiter.waitForEvent(GuildMessageReactionAddEvent::class.java, { check -> check.user == event.author && check.messageId == msg.id }, { action ->
                        // Switch case the reaction emoji
                        when (action.reactionEmote.emoji) {
                            "✅" -> {
                                // Define the embed message
                                val embedBuilder = EmbedBuilder()
                                        .setAuthor("Successfully Purged Channel")
                                        .setColor(event.color)
                                        .setDescription("""**»** Channel: ${channel.name}
                                        **»** Purged By: ${event.author.asMention}
                                        **»** On: ${StringUtils.capitalize(time.dayOfWeek.name)} at ${time.hour}:${time.minute}""".trimIndent())

                                // Create the new channel with the exact same properties
                                event.guild.createTextChannel(channel.name)
                                        .setNSFW(channel.isNSFW)
                                        .setSlowmode(channel.slowmode)
                                        .setTopic(channel.topic)
                                        .setParent(channel.parent)
                                        .setPosition(channel.position).queue { newChannel ->
                                            newChannel.permissionOverrides.forEach { override -> override.permissionHolder?.let { channel.createPermissionOverride(it) }?.queue() }
                                            newChannel.sendMessage(event.author.asMention).queue()
                                            newChannel.sendMessage(embedBuilder.build()).queue()
                                        }

                                // Delete the mentioned channel
                                channel.delete().queue()

                                // Shutdown the event waiter
                                waiter.shutdown()
                            }

                            "❌" -> {
                                // Send cancel message
                                event.sendEmbedReply("❌ Cancelled Channel Purge", "You have cancelled purging ${channel.asMention}, To retry type ${event.prefix}purge channel ${channel.asMention}!")

                                // Shutdown event waiter
                                waiter.shutdown()
                            }
                        }

                    })
                }
            }
        }
    }
}